src/main/java/com/cloudant/http/CookieInterceptor.java

```diff
@@ -0,0 +1,161 @@
+import java.io.UnsupportedEncodingException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ *
+ * Adds cookie authentication support to http requests.
+ *
+ * It does this by adding the cookie header for CouchDB
+ * using request interceptor pipeline in {@link HttpConnection}.
+ *
+ * If a response has a response code of 401, it will fetch a cookie from
+ * the server using provided credentials and tell {@link HttpConnection} to reply
+ * the request by setting {@link HttpConnectionInterceptorContext#replayRequest} property to true.
+ *
+ * If the request to get the cookie for use in future request fails with a 401 status code
+ * (or any status that indicates client error) cookie authentication will not be attempted again.
+ *
+ *
+ */
+public  class CookieInterceptor implements HttpConnectionRequestInterceptor, HttpConnectionResponseInterceptor {
+
+    private final static Logger logger = Logger.getLogger(CookieInterceptor.class.getCanonicalName());
+    final String sessionRequestBody;
+    private String cookie = null;
+    private boolean shouldAttemptCookieRequest = true;
+    private final String username;
+
+    /**
+     * Constructs a cookie interceptor.
+     * @param username The username to use when getting the cookie
+     * @param password The password to use when getting the cookie
+     */
+    public CookieInterceptor(String username, String password){
+        this.sessionRequestBody = String.format("name=%s&password=%s",username,password);
+        this.username = username;
+    }
+
+    @Override
+    public HttpConnectionInterceptorContext interceptRequest(HttpConnectionInterceptorContext context) {
+
+        HttpURLConnection connection = context.connection.getConnection();
+
+        if(shouldAttemptCookieRequest) {
+            if (cookie == null) {
+                cookie = getCookie(connection.getURL());
+            }
+            connection.setRequestProperty("Cookie", cookie);
+        }
+
+        return context;
+    }
+
+    @Override
+    public HttpConnectionInterceptorContext interceptResponse(HttpConnectionInterceptorContext context) {
+        HttpURLConnection connection = context.connection.getConnection();
+        try {
+            if (context.connection.getConnection().getResponseCode() == 401) {
+                //we need to get a new cookie
+                cookie = getCookie(connection.getURL());
+                //don't resend request, failed to get cookie
+                if(cookie != null) {
+                    context.replayRequest = true;
+                } else {
+                    context.replayRequest = false;
+                }
+            }
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, "Failed to get response code from request",e);
+        }
+        return context;
+
+    }
+
+     public String getCookie(URL url){
```

Comment:
I'm sync-android this is private. Why has it been made public? The interceptor itself was designed take care of getting the cookie from the remote server, using it in requests and renewing it etc.